### ISSUE TITLE
Support filtering files by extension

### ISSUE THREAD

**Issue by mmarti** [fastify/fastify-static#145]
## 🚀 Feature Proposal

Support filtering files by extension.

## Motivation

Sometimes you may want to keep your JavaScript or CSS files alongside a component.  For those not using build steps, it'd be great to have the ability to filter what files to serve by pointing `root` at the directory, but only serving certain files based off extension (or more generically could be pattern matching in general).

## Example

Assume we have the following directory structure:

```
src/
├── components/
│   ├── cool-feature
│   │   ├── index.pug
│   │   ├── index.js
│   │   └── index.css
```

We'd register our plugin like so (assume PWD is `src`):

```js
app.register(require('fastify-static'), {
  root: path.join(__dirname, 'components')],
  prefix: '/static',
  extensions: ['.js', '.css'],
});
```

Example of general pattern instead:

```js
app.register(require('fastify-static'), {
  root: path.join(__dirname, 'components')],
  prefix: '/static',
  filter: /\.(?:js|css)$/,
});
```

Examples of access:

```
/static/cool-feature/index.js -> SUCCESS
/static/cool-feature/index.css -> SUCCESS
/static/cool-feature/index.pug -> 404
```

**Comment by mcollina** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by mmarti** [comment]
@mcollina Sure, I should be able to take a first stab.  Quick question, did you have a preference for the implementation?  The most flexible/generic approach would likely be supporting a filter function that would accept the URL/filename and return a boolean.  This would allow users to put as simple or as complex of logic as they want.  Let me know your recommendation and I can work on getting something written.

Thanks.

**Comment by mcollina** [comment]
A filter function would be great.

**Comment by mmarti** [comment]
I just realized you could pass a glob to the `wildcard` option.  If that works the way I think it would, that would actually let you filter (as long as it was based on filename logic).  It would also potentially solve or at least be a workaround for supporting multiple root paths as well, because you could point the root at the shared parent and then use the glob to restrict what child folders should be accessible.

Can you confirm if my understanding of how the wildcard/glob works?

**Comment by mcollina** [comment]
Yes, it should work the way you assume.

**Comment by gajus** [comment]
TLDR use `allowedPath`, e.g.

```ts
allowedPath: (pathName) => {
  if (environmentIsProduction) {
    return !pathName.includes('.map');
  }

  return true;
},
```
